art_internal.hpp

```diff
@@ -12,190 +12,322 @@
 #include "assert.hpp"
+#include "heap.hpp"
 #include "node_type.hpp"
+#include "portability_builtins.hpp"
 
 namespace unodb::detail {
 
 // Forward declarations to use in unodb::db and its siblings
-template <class>
+template <class, class>
 class [[nodiscard]] basic_leaf;
 
-template <class, class>
+template <typename, class, template <class> class>
 class [[nodiscard]] basic_db_leaf_deleter;
 
-// Internal ART key in binary-comparable format
+/// Lexicographic comparison of bytes.
+///
+/// @return -1, 0, or 1 if this key is LT, EQ, or GT the other key.
+[[nodiscard, gnu::pure]] inline int compare(const void *a, const size_t alen,
+                                            const void *b, const size_t blen) {
+  const auto shared_length = std::min(alen, blen);
+  auto ret = std::memcmp(a, b, shared_length);
+  ret = (ret != 0) ? ret : ((alen == blen) ? 0 : (alen < blen ? -1 : 1));
+  return ret;
+}
+
+/// Lexicographic comparison of key_views.
+///
+/// @return -1, 0, or 1 if this key is LT, EQ, or GT the other key.
+[[nodiscard, gnu::pure]] inline int compare(const unodb::key_view a,
+                                            const unodb::key_view b) {
+  return compare(a.data(), a.size_bytes(), b.data(), b.size_bytes());
+}
+
+/// Return the first 64-bits of the encoded key.  This is used by the
+/// prefix compression logic to identify some number of bytes that are
+/// in common between the art_key and an inode having some key_prefix.
+[[nodiscard, gnu::pure]] inline std::uint64_t get_u64(key_view key) noexcept {
+  std::uint64_t u{};  // will hold the first 64-bits.
+  std::memcpy(&u, key.data(), std::min(key.size_bytes(), sizeof(u)));
+  return u;
+}
+
+/// Internal ART key in binary-comparable format.  Application keys may
+/// be simple fixed width types (such as std::uint64_t) or variable
+/// length keys.  For the former, there are convenience methods on db,
+/// olc_db, etc. to convert external keys into the binary compariable
+/// format.  For the latter, the application is responsible for
+/// converting the data (e.g., certain columns in some ordering for a
+/// row of some relation) into the internal binary comparable key
+/// format.  A convenience class (unodb::key_encoder) is offered to
+/// encode data.  The encoding is always well defined and decoding
+/// (unodb::key_decoder) exists for all simple fixed width data types.
+/// Unicode encoding is complex and out of scope - use a quality
+/// library such as ICU to produce appropriate Unicode sort keys for
+/// your application.  Unicode decoding is NOT well defined.
+/// Applications involving database records and Unicode data will
+/// typically store the record identifier in a secondary index (ART) as
+/// the value associated with the key.  Using the record identifier,
+/// the original tuple can be discovered and the original Unicode data
+/// recovered from that tuple.
+//
+/// TODO(thompsonbry) key templating - This never supported anything
```

Comment:
@laurynas-biveinis Just FYI, this is why nothing other than uint64 would work before.

Reply:
OK, now this comment is outdated, right? The part about testing can stay

Reply:
Removed comment.  The issue is covered by the comment on the static assert for art.hpp, etc.